README.md
# consul-c

C binding library for consul client.


debian/changelog
libconsul-c (0.0.1) wazo-dev; urgency=medium

  * Initial release.

 -- Wazo Maintainers <dev+pkg@wazo.community>  Thu, 15 Oct 2020 10:11:00 -0400


client.c
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <stdlib.h>
#include <string.h>

#include <jansson.h>

#include "consul.h"
#include "base64.h"

consul_client_t* consul_client_create(int server_count, const char** servers) {
    int i;
    consul_client_t* client = (consul_client_t*) calloc(1, sizeof(consul_client_t));

    if (server_count < 1)
        return NULL;

    client->server_count = server_count;
    client->servers = (consul_server_t**) calloc(server_count, sizeof(consul_server_t*));
    client->settings.connect_timeout = 5000;
    if (!client->servers)
        return NULL;

    for (i = 0; i < server_count; i++) {
        if ((client->servers[i] = consul_server_create(servers[i])) == NULL) {
            return NULL;
        }
    }

    return client;
}

void consul_client_setup_user(consul_client_t* client, const char *user, const char *password) {
    client->settings.user = strdup(user);
}

void consul_client_destroy(consul_client_t* client) {
    int i = 0;

    if (client->settings.user) {
        free(client->settings.user);
    }

    for (i = 0; i < client->server_count; i++) {
        consul_server_cleanup(client->servers[i]);
    }

    free(client->servers);
}

int consul_parse_lsdir_response(consul_response_t* response, json_t* root) {
    json_t *jstr;
    int count = json_array_size(root);

    response->key_count = count;
    response->keys = calloc(count, sizeof(char*));

    for (int i = 0; i < count; i++){
        jstr = json_array_get(root, i);
        if (jstr == NULL || !json_is_string(jstr)) continue;
        response->keys[i] = (char*) json_string_value(jstr);
    }

    return 0;
}

static int consul_parse_success(consul_response_t* response, json_t* root) {
    json_unpack(root, "b", &response->success);
    return 0;
}

static char *consul_ensure_folder(const char *key) {
    char *folder = strdup(key);
    int size = strlen(key);
    if (!size) {
        return NULL;
    }

    if (folder[size-1] != '/') {
        strcat(folder, "/");
    }

    return folder;
}

int consul_parse_get_response(consul_response_t* response, json_t* root) {
    json_unpack(root, "[{s:i, s:s, s:i, s:s, s:i, s:i}]", "LockIndex", &response->lock_index,
        "Key", &response->key, "Flags", &response->flags, "Value", &response->value,
        "CreateIndex", &response->create_index, "ModifyIndex", &response->modify_index);

    unsigned char *decoded;
    long unsigned decoded_size;

    if (response->value) {
        decoded = base64_decode((const unsigned char*) response->value, strlen(response->value), &decoded_size);

        if (decoded_size) {
            response->value = (char*) decoded;
        }
    }

    return 0;
}

consul_request_t* consul_client_request_create_get(consul_client_t *client, const char *key, int recursive, int keys) {
    CURLU *url = consul_url_create(CONSUL_API_VERSION, CONSUL_KEYS, key);

    if (keys) {
        curl_url_set(url, CURLUPART_QUERY, "keys=true", CURLU_APPENDQUERY);
        curl_url_set(url, CURLUPART_QUERY, "separator=%2F", CURLU_APPENDQUERY);
    }

    if (recursive) {
        curl_url_set(url, CURLUPART_QUERY, "recurse=true", CURLU_APPENDQUERY);
    }

    return consul_client_request_create(client, CONSUL_HTTP_GET, url, NULL);
}

consul_response_t *consul_client_get(consul_client_t *client, const char *key, int recursive, int keys) {
    consul_request_t *req;
    consul_response_t *resp;

    req = consul_client_request_create_get(client, key, recursive, keys);
    resp = consul_cluster_request(client, req);
    if (resp) {
        if (keys) {
            consul_response_parse(resp, consul_parse_get_response);
        } else {
            consul_response_parse(resp, consul_parse_lsdir_response);
        }
    }
    consul_request_cleanup(req);
    return resp;
}

consul_response_t *consul_client_lsdir(consul_client_t *client, const char *key, int recursive) {
    return consul_client_get(client, key, recursive, 1);
}


consul_response_t *consul_client_put(consul_client_t *client, const char *key, char *value) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;

    url = consul_url_create(CONSUL_API_VERSION, CONSUL_KEYS, key);
    req = consul_client_request_create(client, CONSUL_HTTP_PUT, url, value);
    resp = consul_cluster_request(client, req);
    if (resp) {
        consul_response_parse(resp, consul_parse_success);
    }

    curl_url_cleanup(url);
    return resp;
}

consul_response_t *consul_client_delete(consul_client_t *client, const char *key) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;

    url = consul_url_create(CONSUL_API_VERSION, CONSUL_KEYS, key);
    req = consul_client_request_create(client, CONSUL_HTTP_DELETE, url, NULL);
    resp = consul_cluster_request(client, req);
    if (resp) {
        consul_response_parse(resp, consul_parse_success);
    }

    curl_url_cleanup(url);
    return resp;
}

consul_response_t *consul_client_mkdir(consul_client_t *client, const char *key) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;
    char *folder;

    folder = consul_ensure_folder(key);
    url = consul_url_create(CONSUL_API_VERSION, CONSUL_KEYS, folder);
    req = consul_client_request_create(client, CONSUL_HTTP_PUT, url, NULL);
    resp = consul_cluster_request(client, req);
    if (resp) {
        consul_response_parse(resp, consul_parse_success);
    }

    curl_url_cleanup(url);
    free(folder);
    return resp;
}

consul_response_t *consul_client_rmdir(consul_client_t *client, const char *key, int recursive) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;
    char *folder;

    folder = consul_ensure_folder(key);
    url = consul_url_create(CONSUL_API_VERSION, CONSUL_KEYS, folder);
    if (recursive){
        curl_url_set(url, CURLUPART_QUERY, "recursive=true", CURLU_APPENDQUERY);
    }
    req = consul_client_request_create(client, CONSUL_HTTP_DELETE, url, NULL);
    resp = consul_cluster_request(client, req);
    if (resp) {
        consul_response_parse(resp, consul_parse_success);
    }

    curl_url_cleanup(url);
    free(folder);
    return resp;
}


server.c
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */


#include <stdlib.h>

#include "consul.h"

consul_server_t* consul_server_create(const char *url) {
    consul_server_t *server;
    CURLU* curlu;
    CURLMcode rc;

    server = (consul_server_t*) calloc(1, sizeof(consul_server_t));
    if (server) {
        curlu = curl_url();

        if ((rc = curl_url_set(curlu, CURLUPART_URL, url, 0)) != CURLM_OK) {
            return NULL;
        }

        curl_url_get(curlu, CURLUPART_SCHEME, &server->scheme, 0);
        curl_url_get(curlu, CURLUPART_HOST, &server->host, 0);
        curl_url_get(curlu, CURLUPART_PORT, &server->port, 0);

        curl_url_cleanup(curlu);
    }

    return server;
}

void consul_server_cleanup(consul_server_t* server) {
    if (server->host)
        curl_free(server->host);
    if (server->port)
        curl_free(server->port);
    if (server->scheme)
        curl_free(server->scheme);
}

request.c
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <stdlib.h>
#include <string.h>

#include "consul.h"

static const char *http_method[] = {
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "HEAD",

    "OPTION"
};

CURLU* consul_url_create(int version, enum CONSUL_API_TYPE type, const char *path) {
    CURLU *url;
    CURLMcode rc;
    char url_path[PATH_MAX];
    const char *prefix;

    switch (type) {
    case CONSUL_KEYS: prefix = "kv"; break;
    case CONSUL_SERVICES: prefix = "health/service"; break;
    case CONSUL_AGENT_REGISTER: prefix = "agent/service/register"; break;
    case CONSUL_AGENT_DEREGISTER: prefix = "agent/service/deregister"; break;
    case CONSUL_AGENT_SET_MAINTENANCE: prefix = "agent/service/maintenance"; break;
    }

    url = curl_url();
    if (path && path[0] != '\0') {
        snprintf(url_path, PATH_MAX, "/v%d/%s/%s", version, prefix, path);
    } else {
        snprintf(url_path, PATH_MAX, "/v%d/%s", version, prefix);
    }
    if ((rc = curl_url_set(url, CURLUPART_PATH , url_path, 0)) != CURLM_OK) {
        curl_url_cleanup(url);
        return NULL;
    };

    return url;
}

size_t consul_header_callback(char *buffer, size_t size,
                              size_t nitems, void *userdata)
{
    int status_code;
    char *sep;
    char *saveptr;
    char* key;
    char *value;
    char *suffix;
    consul_response_t* resp;

    int count = nitems * size;

    resp = (consul_response_t*) userdata;

    if (!resp->err) {
        resp->err = (consul_error_t*) calloc(1, sizeof(consul_error_t));

        buffer = strtok_r(buffer, "\r\n", &saveptr);
        if (!buffer) {
            resp->err->ecode = ERROR_INVALID_RESPONSE;
            return 0;
        }

        const char* http_version = "HTTP/1.1 ";
        size_t http_size = strlen(http_version);
        if (strncmp(buffer, http_version, http_size)) {
            resp->err->ecode = ERROR_INVALID_RESPONSE;
            return 0;
        }

        buffer += http_size;

        if (sscanf(buffer, "%3d", &status_code) != 1) {
            resp->err->ecode = ERROR_INVALID_RESPONSE;
            return 0;
        }

        buffer += 3;
        if (buffer[0] != '\0') {
            resp->err->message = strdup(buffer+1);
        }

        resp->err->ecode = status_code;
    } else {
        if ((sep = index(buffer, ':'))) {
            *sep = '\0';
            key = buffer;
            buffer = sep + 1;
            while (*buffer != '\0' && *buffer == ' ') buffer++;
            if ((suffix = strstr(buffer, "\r\n"))) {
                *suffix = '\0';
                value = buffer;
                if (!(strcmp(key, "X-Consul-Index"))) {
                    resp->modify_index = atoi(value);
                } else if (!strcmp(key, "Transfer-Encoding") && !strcmp(value, "chunked")) {
                    resp->chunked = 1;
                }
            }
        }
    }

    return count;
}

size_t consul_body_callback(void *contents, size_t length, size_t nmemb, void *userp) {
    consul_response_t* resp = (consul_response_t*) userp;
    size_t size = length * nmemb;
    if (resp->data && resp->chunked) {
        size_t previous_size = resp->data_length;
        resp->data_length += size;
        resp->data = realloc(resp->data, resp->data_length);
        strncpy(resp->data+previous_size, (const char*) contents, size);
    }
    else {
        resp->data = malloc(size);
        strncpy(resp->data, contents, size);
        resp->data_length = size;
    }
    return size;
}

int consul_response_is_success(consul_response_t *response) {
    return response->err && (response->err->ecode >= 200) && (response->err->ecode < 300);
}

consul_response_t* consul_response_create() {
    return (consul_response_t*) calloc(1, sizeof(consul_response_t));
}

void consul_request_setopt(consul_request_t* req, consul_response_t* resp, CURL *curl) {
    curl_easy_reset(curl);

    curl_easy_setopt(curl, CURLOPT_CURLU, req->url);
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, http_method[req->method]);
    if (req->client->settings.user) {
        curl_easy_setopt(curl, CURLOPT_USERNAME, req->client->settings.user);
    }
    if (req->client->settings.password) {
        curl_easy_setopt(curl, CURLOPT_PASSWORD, req->client->settings.password);
    }

    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);

    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, consul_header_callback);
    curl_easy_setopt(curl, CURLOPT_HEADERDATA, resp);

    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, consul_body_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, resp);

    if (req->data) {
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->data);
    }

    #if LIBCURL_VERSION_NUM >= 0x071900
    curl_easy_setopt(curl, CURLOPT_TCP_KEEPALIVE, 1L);
    curl_easy_setopt(curl, CURLOPT_TCP_KEEPINTVL, 1L); /*the same as go-etcd*/
    #endif
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "consul-client");
    curl_easy_setopt(curl, CURLOPT_POSTREDIR, 3L);     /*post after redirecting*/

    curl_easy_setopt(curl, CURLOPT_VERBOSE, req->client->settings.verbose);
    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, req->client->settings.connect_timeout);
}

consul_response_t* consul_request_send(consul_request_t* req) {
    CURLMcode rc;
    consul_response_t* resp = NULL;

    resp = (consul_response_t*) calloc(1, sizeof(consul_response_t));

    consul_request_setopt(req, resp, req->curl);

    rc = curl_easy_perform(req->curl);

    if (rc != CURLM_OK) {
        if (resp->err == NULL) {
            resp->err = (consul_error_t*) calloc(1, sizeof(consul_error_t));
            resp->err->ecode = ERROR_REQUEST_FAILED;
            resp->err->message = strdup(curl_easy_strerror(rc));
        }
    }

    return resp;
}

void consul_response_reset(consul_response_t* response) {
    if (response->err) {
        if (response->err->message) {
            free(response->err->message);
        }
        free(response->err);
        response->err = NULL;
    }

    if (response->value) {
        free(response->value);
        response->value = NULL;
    }

    if (response->key) {
        free(response->key);
        response->key = NULL;
    }

    if (response->keys) {
        for (int i = 0; i < response->key_count; i++) {
            free(response->keys[i]);
        }
        free(response->keys);
        response->key_count = 0;
        response->keys = NULL;
    }

    if (response->data) {
        free(response->data);
        response->data = NULL;
    }

    memset(response, 0, sizeof(consul_response_t));
}

void consul_response_cleanup(consul_response_t* response) {
    consul_response_reset(response);
    free(response);
}

consul_request_t* consul_client_request_create(consul_client_t* client, enum HTTP_METHOD method, CURLU* url, const char *data) {
    consul_request_t* req = (consul_request_t*) calloc(1, sizeof(consul_request_t));
    req->curl = curl_easy_init();
    req->method = method;
    req->url = url;
    req->client = client;
    req->data = data;
    return req;
}

int consul_response_parse(consul_response_t *response, consul_response_parser parser) {
    json_t* root;
    json_error_t err;

    if (response->data) {
        root = json_loadb(response->data, response->data_length, JSON_DECODE_ANY, &err);
        if (!root) {
            printf("failed to load json (%s), %s, line %d, column %d, position %d!\n", response->data, err.text, err.line, err.column, err.position);
            return -1;
        }
        return parser(response, root);
    }

    return -1;
}

void consul_request_set_server(consul_request_t* request, consul_server_t* server) {
    curl_url_set(request->url, CURLUPART_SCHEME, server->scheme, 0);
    curl_url_set(request->url, CURLUPART_HOST, server->host, 0);
    curl_url_set(request->url, CURLUPART_PORT, server->port, 0);
}

void consul_request_cleanup(consul_request_t* request) {
    curl_easy_cleanup(request->curl);
    if (request->data)
        free((void*) request->data);
}

consul_response_t* consul_cluster_request(consul_client_t* client, consul_request_t* req) {
    consul_response_t* resp;
    int i;

    for (i = 0; i < client->server_count; i++) {
        consul_request_set_server(req, client->servers[client->leader]);

        resp = consul_request_send(req);

        if (!resp || !resp->err || resp->err->ecode == ERROR_REQUEST_FAILED) {
            client->leader = (client->leader + 1) % client->server_count;
            continue;
        } else {
            break;
        }
    }

    return resp;
}

service.c
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include "consul.h"

#include <jansson.h>

static json_t* strings_to_json_array(const char** strings) {
    json_t* array = json_array();

    while (*strings) {
        json_array_append(array, json_string(*strings));
        strings++;
    }

    return array;
}

static json_t* strings_to_json_dict(const char** strings) {
    json_t* obj = json_object();

    while (*strings) {
        json_object_set(obj, *strings, json_string(*(strings+1)));
        strings += 2;
    }

    return obj;
}

static json_t* check_to_json_dict(consul_check_t *check) {
    char interval[32];
    snprintf(interval, 32, "%ds", check->interval);
    return json_pack("{s:s?,s:s}", "HTTP", check->http, "Interval", interval);
}

static json_t* checks_to_json_array(consul_check_t **checks) {
    json_t* array = json_array();

    while (*checks) {
        json_array_append(array, check_to_json_dict(*checks));
        checks++;
    }

    return array;
}

consul_response_t *consul_service_register(consul_client_t* client, consul_service_t* service) {
    consul_request_t *req;
    consul_response_t *resp;
    json_t* obj;
    CURLU *url;
    char *dump;

    obj = json_pack("{s:s?,s:s?,s:s?,s:i,s:o*,s:o*,s:o*}",
                    "ID", service->id,
                    "Name", service->name,
                    "Address", service->address,
                    "Port", service->port,
                    "Tags", service->tags ? strings_to_json_array(service->tags) : NULL,
                    "Meta", service->meta ? strings_to_json_dict(service->meta) : NULL,
                    "Checks", service->checks ? checks_to_json_array(service->checks) : NULL);

    dump = json_dumps(obj, 0);
    url = consul_url_create(CONSUL_API_VERSION, CONSUL_AGENT_REGISTER, "");
    req = consul_client_request_create(client, CONSUL_HTTP_PUT, url, dump);
    resp = consul_cluster_request(client, req);
    consul_request_cleanup(req);
    return resp;
}

consul_response_t *consul_service_deregister(consul_client_t* client, const char *service_id) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;

    url = consul_url_create(CONSUL_API_VERSION, CONSUL_AGENT_DEREGISTER, service_id);
    req = consul_client_request_create(client, CONSUL_HTTP_PUT, url, NULL);
    resp = consul_cluster_request(client, req);
    consul_request_cleanup(req);
    return resp;
}

consul_response_t *consul_service_set_maintenance(consul_client_t* client, const char *service_id, int state, const char *reason) {
    consul_request_t *req;
    consul_response_t *resp;
    CURLU *url;
    char reason_param[4096];

    url = consul_url_create(CONSUL_API_VERSION, CONSUL_AGENT_SET_MAINTENANCE, service_id);
    if (state) {
        curl_url_set(url, CURLUPART_QUERY, "enable=true", CURLU_APPENDQUERY);
    } else {
        curl_url_set(url, CURLUPART_QUERY, "enable=false", CURLU_APPENDQUERY);
    }
    if (reason) {
        snprintf(reason_param, 4096, "reason=%s", reason);
        curl_url_set(url, CURLUPART_QUERY, reason_param, CURLU_APPENDQUERY);
    }

    req = consul_client_request_create(client, CONSUL_HTTP_PUT, url, NULL);
    resp = consul_cluster_request(client, req);
    consul_request_cleanup(req);
    return resp;
}


watcher.c
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <stdlib.h>
#include <string.h>

#include "consul.h"

static CURLU* consul_watcher_build_url(consul_client_t *client, consul_watcher_t *watcher) {
    CURLU* url = consul_url_create(CONSUL_API_VERSION, watcher->type, watcher->key);
    if (!url)
        return NULL;

    char wait_param[64];
    snprintf(wait_param, 64, "wait=%ds", watcher->wait_timeout);
    curl_url_set(url, CURLUPART_QUERY, wait_param, CURLU_APPENDQUERY);

    if (watcher->index) {
        char index_param[64];
        snprintf(index_param, 64, "index=%d", watcher->index);
        curl_url_set(url, CURLUPART_QUERY, index_param, CURLU_APPENDQUERY);
    }

    if (watcher->recursive) {
        curl_url_set(url, CURLUPART_QUERY, "recurse=true", CURLU_APPENDQUERY);
    }

    if (watcher->keys) {
        curl_url_set(url, CURLUPART_QUERY, "keys=true", CURLU_APPENDQUERY);
    }

    return url;
}

consul_watcher_t* consul_watcher_create(consul_client_t *client, enum CONSUL_API_TYPE type, const char *key, int recursive, int keys, int initial, consul_watcher_callback_t cb, void *userdata, consul_response_parser parser, int wait_timeout) {
    consul_watcher_t* watch = (consul_watcher_t*) calloc(1, sizeof(consul_watcher_t));

    if (watch) {
        watch->client = client;
        watch->type = type;
        watch->key = key;
        watch->curl = curl_easy_init();
        watch->recursive = recursive;
        watch->keys = keys;
        watch->callback = cb;
        watch->wait_timeout = wait_timeout;
        watch->userdata = userdata;
        watch->parser = parser;
        watch->index = 1;

        watch->request = consul_client_request_create_get(client, key, recursive, keys);
        watch->response = consul_cluster_request(client, watch->request);
        if (watch->response) {
            consul_response_parse(watch->response, parser);
            if (initial) {
                cb(watch->response, userdata);
            }
            watch->index = watch->response->modify_index;
        }

        consul_watcher_reset(watch);
    }

    return watch;
}

int consul_watcher_reset(consul_watcher_t *watcher) {
    CURL* curl = watcher->curl;

    if (watcher->request->url)
        free(watcher->request->url);

    watcher->request->url = consul_watcher_build_url(watcher->client, watcher);

    consul_response_reset(watcher->response);

    consul_request_set_server(watcher->request, watcher->client->servers[watcher->client->leader]);

    consul_request_setopt(watcher->request, watcher->response, curl);

    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
    curl_easy_setopt(curl, CURLOPT_PRIVATE, watcher);

    return 0;
}

static int consul_check_watchers(consul_client_t *cli, CURLM *mcurl) {
    uint index = 0;
    int error = 0;
    int ignore;
    CURLMsg *msg;
    CURL *curl;
    consul_watcher_t *watcher;
    consul_response_t *resp;

    while ((msg = curl_multi_info_read(mcurl, &ignore)) != NULL) {
        if (msg->msg == CURLMSG_DONE) {
            curl = msg->easy_handle;
            curl_easy_getinfo(curl, CURLINFO_PRIVATE, &watcher);

            resp = watcher->response;
            index = watcher->index;
            if (msg->data.result != CURLE_OK) {
                if (watcher->attempts > 1) {
                    cli->leader = (cli->leader + 1) % cli->server_count;
                    watcher->attempts--;
                    error = ERROR_REQUEST_FAILED;
                } else if (watcher->attempts == 1) {
                    watcher->attempts--;
                    error = resp->err->ecode = ERROR_CLUSTER_FAILED;
                }
            }

            if (resp->err && resp->err->ecode >= 100 && resp->err->ecode < 500) {
                consul_response_parse(resp, watcher->parser);

                if (watcher->callback && resp->modify_index > index) {
                    error = watcher->callback(resp, watcher->userdata);
                }

                watcher->index = resp->modify_index;
            }

            if (!error) {
                curl_multi_remove_handle(mcurl, curl);
                consul_watcher_reset(watcher);
                curl_multi_add_handle(mcurl, watcher->curl);
                continue;
            }

            curl_multi_remove_handle(mcurl, curl);
            consul_watcher_destroy(watcher);
        }
    }

    return error;
}

int consul_multi_watch(consul_client_t *client, consul_watcher_t **watchers) {
    int              maxfd, left, error;
    long             backoff, backoff_max;
    consul_watcher_t *watcher;
    CURLM            *mcurl;
    CURLMcode        mc;

    struct timeval tv;

    mcurl = curl_multi_init();
    while ((watcher = *watchers++)) {
        curl_easy_setopt(watcher->curl, CURLOPT_PRIVATE, watcher);
        curl_multi_add_handle(mcurl, watcher->curl);
    }

    backoff = 100;
    backoff_max = 1000;

    for(;;) {
        mc = curl_multi_perform(mcurl, &left);
        if (mc == CURLM_OK) {
            mc = curl_multi_wait(mcurl, NULL, 0, 1000, &maxfd);
        }
        if (mc != CURLM_OK) {
            fprintf(stderr, "error curl_multi_perform: %s\n", curl_multi_strerror(mc));
            break;
        }

        error = consul_check_watchers(client, mcurl);
        if (error || left == 0) {
            if (backoff < backoff_max) {
                backoff = 2 * backoff;
            } else {
                 backoff = backoff_max;
            }

            tv.tv_sec = backoff/1000;
            tv.tv_usec = (backoff%1000) * 1000;
            select(1, 0, 0, 0, &tv);
        }
    }

    curl_multi_cleanup(mcurl);
    return mc != CURLM_OK;
}

void consul_watcher_destroy(consul_watcher_t* watch) {
    if (watch->curl)
        curl_easy_cleanup(watch->curl);
    if (watch->request)
        consul_request_cleanup(watch->request);
    if (watch->response)
        consul_response_cleanup(watch->response);
    free(watch);
}


consul.h
/*
 * Copyright 2020 The Wazo Authors  (see the AUTHORS file)
 * SPDX-License-Identifier: BSD-2-Clause
 */


#include <curl/curl.h>
#include <jansson.h>

#define CONSUL_API_VERSION 1
#define MAX_WATCHERS       32
#define MAX_CONSUL_SERVERS 15

#define TOO_MANY_SERVERS -1
#define OUT_OF_MEMORY    -2

#define ERROR_REQUEST_FAILED   1001
#define ERROR_INVALID_RESPONSE 1002
#define ERROR_CLUSTER_FAILED   1003

enum CONSUL_API_TYPE {
    CONSUL_KEYS,
    CONSUL_SERVICES,
    CONSUL_AGENT_REGISTER,
    CONSUL_AGENT_DEREGISTER,
    CONSUL_AGENT_SET_MAINTENANCE
};

enum HTTP_METHOD {
    CONSUL_HTTP_GET,
    CONSUL_HTTP_POST,
    CONSUL_HTTP_PUT,
    CONSUL_HTTP_DELETE,
    CONSUL_HTTP_HEAD,
    CONSUL_HTTP_OPTION
};

struct consul_response_t;
struct consul_request_t;

typedef int (*consul_watcher_callback_t) (struct consul_response_t* response, void *userdata);

typedef int (*consul_response_parser) (struct consul_response_t *response, json_t* root);

typedef struct consul_watcher_t {
    struct consul_client_t*   client;
    CURLM*                    curl;
    int                       server_count;
    int                       recursive;
    int                       index;
    int                       keys;
    enum CONSUL_API_TYPE      type;
    const char*               key;
    int                       wait_timeout;
    int                       attempts;
    consul_watcher_callback_t callback;
    void*                     userdata;
    consul_response_parser    parser;
    struct consul_response_t* response;
    struct consul_request_t*  request;
} consul_watcher_t;

typedef struct consul_server_t {
    char *host;
    char *port;
    char *scheme;
} consul_server_t;

typedef struct consul_client_t {
    consul_server_t** servers;
    int    server_count;
    int leader;
    consul_watcher_t* watchers[MAX_WATCHERS];
    struct {
        int      verbose;
        uint     ttl;
        uint     connect_timeout;
        char*    user;
        char*    password;
        char*    scheme;
    } settings;
} consul_client_t;

typedef struct consul_error_t {
    int ecode;
    char *message;
    char *cause;
} consul_error_t;

typedef struct consul_response_t {
    consul_error_t* err;
    uint index;
    int known_leader;
    int last_contact;
    char *data;
    int data_length;
    int chunked;
    uint modify_index;
    struct {
        int key_count;
        char** keys;
    };
    struct {
        int lock_index;
        uint create_index;
        char* key;
        char *value;
        int flags;
    };
    int success;
} consul_response_t;

typedef struct consul_check_t {
    const char *http;
    int interval;
} consul_check_t;

typedef struct consul_service_t {
    struct {
        char* id;
        char *name;
        char* node;
        char* address;
        int port;
        char **tags;
        char **meta;
    } Node;

    const char* id;
    const char *name;
    const char* node;
    const char* address;
    int port;
    const char **tags;
    const char **meta;
    consul_check_t **checks;
} consul_service_t;

typedef struct consul_request_t {
    enum HTTP_METHOD method;
    CURL *curl;
    CURLU* url;
    const char *data;
    consul_client_t* client;
} consul_request_t;

CURLU* consul_url_create(int version, enum CONSUL_API_TYPE type, const char *path);

consul_server_t* consul_server_create(const char *url);
void consul_server_cleanup(consul_server_t* server);

consul_client_t* consul_client_create(int server_count, const char** servers);
void consul_client_setup_user(consul_client_t* client, const char *user, const char *password);
void consul_client_destroy(consul_client_t* client);
consul_response_t *consul_client_lsdir(consul_client_t *cli, const char *key, int recursive);
consul_response_t* consul_cluster_request(consul_client_t* client, consul_request_t* req);
consul_response_t *consul_client_put(consul_client_t *client, const char *key, char *value);
consul_response_t *consul_client_get(consul_client_t *client, const char *key, int recursive, int keys);
consul_response_t *consul_client_delete(consul_client_t *client, const char *key);
consul_response_t *consul_client_mkdir(consul_client_t *client, const char *key);
consul_response_t *consul_client_rmdir(consul_client_t *client, const char *key, int recursive);
int consul_multi_watch(consul_client_t *client, consul_watcher_t **watchers);

consul_watcher_t* consul_watcher_create(consul_client_t *client, enum CONSUL_API_TYPE type, const char *key, int recursive, int keys, int initial, consul_watcher_callback_t cb, void *userdata, consul_response_parser parser, int wait_timeout);
int consul_watcher_reset(consul_watcher_t *watcher);
void consul_watcher_destroy(consul_watcher_t* watch);

size_t consul_header_callback(char *buffer, size_t size, size_t nitems, void *userdata);
size_t consul_body_callback(void *contents, size_t length, size_t nmemb, void *userp);

consul_request_t* consul_client_request_create(consul_client_t* client, enum HTTP_METHOD method, CURLU* url, const char *data);
consul_request_t* consul_client_request_create_get(consul_client_t *client, const char *key, int recursive, int keys);
void consul_request_set_server(consul_request_t* request, consul_server_t* server);
consul_response_t* consul_request_send(consul_request_t* req);
void consul_request_setopt(consul_request_t* req, consul_response_t* resp, CURL *curl);
void consul_request_cleanup(consul_request_t* request);

consul_response_t* consul_response_create(void);
int consul_response_is_success(consul_response_t *response);
void consul_response_reset(consul_response_t* response);
void consul_response_cleanup(consul_response_t* response);
int consul_response_parse(consul_response_t *response, consul_response_parser parser);

int consul_parse_get_response(consul_response_t* response, json_t* root);
int consul_parse_lsdir_response(consul_response_t* response, json_t* root);

consul_response_t *consul_service_register(consul_client_t* client, consul_service_t* service);
consul_response_t *consul_service_deregister(consul_client_t* client, const char *service_id);
consul_response_t *consul_service_set_maintenance(consul_client_t* client, const char *service_id, int state, const char *reason);
